homeassistant/components/homekit_controller/humidifier.py

```diff
@@ -0,0 +1,351 @@
+        data = {
+            ATTR_MIN_HUMIDITY: self.min_humidity,
+            ATTR_MAX_HUMIDITY: self.max_humidity,
+            ATTR_AVAILABLE_MODES: self.available_modes,
+        }
+
+        return data
+
+    @property
+    def state_attributes(self) -> Dict[str, Any]:
+        """Return the optional state attributes."""
+        data = {
+            ATTR_MODE: self.mode,
+            ATTR_HUMIDITY: self.target_humidity,
+        }
+
+        return data
+
+    @property
+    def target_humidity(self) -> Optional[int]:
+        """Return the humidity we try to reach."""
+        return self.service.value(
+            CharacteristicsTypes.RELATIVE_HUMIDITY_DEHUMIDIFIER_THRESHOLD
+        )
+
+    @property
+    def mode(self) -> Optional[str]:
+        """Return the current mode, e.g., home, auto, baby.
+
+        Requires SUPPORT_MODES.
+        """
+        mode = self.service.value(
+            CharacteristicsTypes.CURRENT_HUMIDIFIER_DEHUMIDIFIER_STATE
+        )
+        return MODE_AUTO if mode == 1 else MODE_NORMAL
+
+    @property
+    def available_modes(self) -> Optional[List[str]]:
+        """Return a list of available modes.
+
+        Requires SUPPORT_MODES.
+        """
+        available_modes = [
+            MODE_NORMAL,
+            MODE_AUTO,
+        ]
+
+        return available_modes
+
+    async def async_set_humidity(self, humidity: int) -> None:
+        """Set new target humidity."""
+        await self.async_put_characteristics(
+            {CharacteristicsTypes.RELATIVE_HUMIDITY_DEHUMIDIFIER_THRESHOLD: humidity}
+        )
+
+    async def async_set_mode(self, mode: str) -> None:
+        """Set new mode."""
+        if mode == MODE_AUTO:
+            await self.async_put_characteristics(
+                {
+                    CharacteristicsTypes.TARGET_HUMIDIFIER_DEHUMIDIFIER_STATE: 0,
+                    CharacteristicsTypes.ACTIVE: True,
+                }
+            )
+        else:
+            await self.async_put_characteristics({CharacteristicsTypes.ACTIVE: False})
+
+    @property
+    def min_humidity(self) -> int:
+        """Return the minimum humidity."""
+        return self.service[
+            CharacteristicsTypes.RELATIVE_HUMIDITY_DEHUMIDIFIER_THRESHOLD
+        ].minValue
+
+    @property
+    def max_humidity(self) -> int:
+        """Return the maximum humidity."""
+        return self.service[
+            CharacteristicsTypes.RELATIVE_HUMIDITY_DEHUMIDIFIER_THRESHOLD
+        ].maxValue
+
+    @property
+    def unique_id(self) -> str:
+        """Return the ID of this device."""
+        serial = self.accessory_info.value(CharacteristicsTypes.SERIAL_NUMBER)
+        return f"homekit-{serial}-{self._iid}-{self.device_class}"
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Homekit humidifer."""
+    hkid = config_entry.data["AccessoryPairingID"]
+    conn = hass.data[KNOWN_DEVICES][hkid]
+
+    def get_accessory(conn, aid):
```

Comment:
I think we'll be able to do better here when I have refactor the async_setup_entry code to pass around a aiohomekit `Service` object instead of a dict, but I think this is fine until I do that refactor.